README.md
X-editable for Yii
======================

Bundle of widgets and server-side component for creating editable elements in [Yii](http://www.yiiframework.com) application.
This extension comes instead of previous [yii-bootstrap-editable](http://www.yiiframework.com/extension/yii-bootstrap-editable) that was upgraded to [X-editable](http://vitalets.github.com/x-editable) library.  
Main changes are:

* support of several libraries: Twitter Bootstrap, jQuery UI and pure jQuery
* popup and inline modes. You can toggle it from config without changing code
* update of related models

##Demo & Documentation
Please see **[x-editable.demopage.ru](http://x-editable.demopage.ru)**

##Issues
Feel free to open issue if you find bug or have an idea.
Thanks for your feedback!

##Contribution
Submit your ideas/solutions to `dev` branch. Thank you!

## License
Copyright (c) 2012 Vitaliy Potapov  
Licensed under the MIT licenses.


CHANGELOG.txt
X-editable-yii changelog
============================

Version 1.3.2  Dec 23, 2014
----------------------------
[enh #65] treat `params` as php expression for EditableColumn (vitalets)
[bug #116]  Memory leak in EditableColumn (Julian-B90)


Version 1.3.1  Aug 17, 2013
----------------------------
[enh] conversion nl2br no more needed for textarea (vitalets)
[enh] updated to x-editable 1.4.6 (vitalets)
[enh #63] allow usage with EDataTables (vitalets, faravaghi)
[bug #66] date & datetime are not processed if passed as timestamp string (vitalets)
[enh #19] i18n support for "clear" in date & datetime (vitalets)
[enh #40] evaluate htmlOptions for EditableColumn (horizons2, vitalets)
[enh #57] added support for CFormModel (vitalets)
[enh #55] support of inline functions in column value (vitalets)


Version 1.3.0  Jun 29, 2013
----------------------------
[enh #43] support for non-scalar virtual attributes (pavshuk)
[enh] update x-editable js to 1.4.5 (vitalets)
[enh] added mongoDB support (vitalets)
[enh #48] scenario is taken dynamically from model and used in EditableSaver (vitalets)
[enh #46] add `liveTarget` option to support live update in CListView (vitalets)


Version 1.2.0  May 29, 2013
----------------------------
[enh #7] column sort icons for bootstrap (vitalets)
[bug #41] if all columns have `apply=false` grid becomes un-editable after update (horizons2)
[enh #39] support php < 5.3 (vitalets)
[enh #37] added Editable::source method (vitalets)
[enh #38] CSRF support out of box (horizons2)
[enh #35] support of dates as timestamp (andersonamuller)
[enh] support of virtual_fields in EditableColumn (vitalets)


Version 1.1.0  May 18, 2013
----------------------------
[enh #33] allow `apply` as a PHP expression in EditableColumn (oinegue)
[enh #22] allow CArrayDataProvider for EditableColumn (ciarand)
[enh] added bootstrap datetime support (vitalets)
[enh #5] source as array is always data, not Yii route (vitalets)
[enh #9] support for namespaced models (vitalets)
[enh] updated to x-editable 1.4.4: added combodate and select2 (vitalets)
[enh #16] EditableDetailView defaults + french translation (mdeweerd) 
[bug] If apply = true, attribute can be unsafe (vitalets)
[enh] Added composer support (vitalets)
[chg] EditableDetailView emptytext no more taken from nullDisplay option (vitalets)
[bug] Disable loading Yii's css for EditableDetailView in bootstrap mode (vitalets)
[enh #6] Added support of composite primary keys (vitalets)
[enh] Added Italian messages (vitalets)


Version 1.0.0  Dec 26, 2012
----------------------------
Initial release.

EditableColumn.php
<?php
/**
 * EditableColumn class file.
 *
 * @author Vitaliy Potapov <noginsk@rambler.ru>
 * @link https://github.com/vitalets/x-editable-yii
 * @copyright Copyright &copy; Vitaliy Potapov 2012
 * @version 1.3.2
 */

Yii::import('editable.EditableField');
Yii::import('bootstrap.widgets.TbDataColumn');

/**
 * EditableColumn widget makes editable one column in CGridView.
 *
 * @package widgets
 */
class EditableColumn extends TbDataColumn
{
    /**
    * @var array editable config options.
    * @see EditableField config
    */
    public $editable = array();

    public function init()
    {
        if (!$this->name) {
            throw new CException('You should provide name for EditableColumn');
        }

        parent::init();

        //need to attach ajaxUpdate handler to refresh editables on pagination and sort
        Editable::attachAjaxUpdateEvent($this->grid);
    }
   
    //protected was removed due to https://github.com/vitalets/x-editable-yii/issues/63
    public function renderDataCellContent($row, $data)
    {
        $isModel = $data instanceOf CModel;
        
        if($isModel) {
            $widgetClass = 'EditableField';
            $options = array(
                'model'        => $data,
                'attribute'    => empty($this->editable['attribute']) ? $this->name : $this->editable['attribute'],
            );
            
            //if value defined in column config --> we should evaluate it 
            //and pass to widget via `text` option: set flag `passText` = true 
            $passText = !empty($this->value);     
        } else {
            $widgetClass = 'Editable';
            $options = array(
                'pk'           => $data[$this->grid->dataProvider->keyField],
                'name'         => empty($this->editable['name']) ? $this->name : $this->editable['name'],
            );
            
            $passText = true;
            //if autotext will be applied, do not pass `text` option (pass `value` instead)
            if(empty($this->value) && Editable::isAutotext($this->editable, isset($this->editable['type']) ? $this->editable['type'] : '')) {
               $options['value'] = $data[$this->name]; 
               $passText = false;
            } 
        }
        
        //for live update
        $options['liveTarget'] = $this->grid->id;
        
        $options = CMap::mergeArray($this->editable, $options);

        //if value defined for column --> use it as element text
        if($passText) {
            ob_start();
            parent::renderDataCellContent($row, $data);
            $text = ob_get_clean();
            $options['text'] = $text;
            $options['encode'] = false;
        }
        
        //apply may be a string expression, see https://github.com/vitalets/x-editable-yii/issues/33
        if (isset($options['apply']) && is_string($options['apply'])) {
            $options['apply'] = $this->evaluateExpression($options['apply'], array('data'=>$data, 'row'=>$row));
        }

        //evaluate htmlOptions inside editable config as they can depend on $data
        //see https://github.com/vitalets/x-editable-yii/issues/40
        if (isset($options['htmlOptions']) && is_array($options['htmlOptions'])) {
            foreach($options['htmlOptions'] as $k => $v) {
                if(is_string($v) && (strpos($v, '$data') !== false || strpos($v, '$row') !== false)) {
                    $options['htmlOptions'][$k] = $this->evaluateExpression($v, array('data'=>$data, 'row'=>$row));
                }
            }
        }
        
        //evaluate `params` as they can depend on $data
        //see https://github.com/vitalets/x-editable-yii/issues/65
        if (isset($options['params']) && is_array($options['params'])) {
            foreach($options['params'] as $k => $v) {
                if(is_string($v) && (strpos($v, '$data') !== false || strpos($v, '$row') !== false)) {
                    $options['params'][$k] = $this->evaluateExpression($v, array('data'=>$data, 'row'=>$row));
                }
            }
        }           
        
        $this->grid->controller->widget($widgetClass, $options);
    }
    
    /*
    Require this overwrite to show bootstrap sort icons
    */
    protected function renderHeaderCellContent()
    {
        if(yii::app()->editable->form != EditableConfig::FORM_BOOTSTRAP) {
            parent::renderHeaderCellContent();
            return;
        }
        
        if ($this->grid->enableSorting && $this->sortable && $this->name !== null)
        {
            $sort = $this->grid->dataProvider->getSort();
            $label = isset($this->header) ? $this->header : $sort->resolveLabel($this->name);

            if ($sort->resolveAttribute($this->name) !== false)
                $label .= '<span class="caret"></span>';

            echo $sort->link($this->name, $label, array('class'=>'sort-link'));
        }
        else
        {
            if ($this->name !== null && $this->header === null)
            {
                if ($this->grid->dataProvider instanceof CActiveDataProvider)
                    echo CHtml::encode($this->grid->dataProvider->model->getAttributeLabel($this->name));
                else
                    echo CHtml::encode($this->name);
            }
            else
                parent::renderHeaderCellContent();
        }
    } 
    
    /*
    Require this overwrite to show bootstrap filter field
    */    
    public function renderFilterCell()
    {
        if(yii::app()->editable->form != EditableConfig::FORM_BOOTSTRAP) {
            parent::renderFilterCell();
            return;
        }
                
        echo '<td><div class="filter-container">';
        $this->renderFilterCellContent();
        echo '</div></td>';
    }       
}
